### PR TITLE
Corrections to WaterToAirHeatPump (WSHP) water flow control

### PR DESCRIPTION
**Pull request by rraustad**
## Pull request overview

Corrections were made to the water flow rate and temperature calculations for ZoneHVAC:WaterToAirHeatPump and AirLoopHVAC:UnitaryHeatPump:WaterToAir with Heat Pump Coil Water Flow Mode = Cycling to prevent temperature out of range warnings when calculating fluid specific heat. Fixes defect #4964. Diff's are expected in WSHP example files.
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
### Review Checklist

This will not be exhaustively relevant to every PR.
- [x] Code style (parentheses padding, variable names)
- [x] Functional code review (it has to work!)
- [x] If defect, results of running current develop vs this branch should exhibit the fix
- [x] CI status: all green or justified
- [x] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [x] Argument types
  - [x] If any virtual classes, ensure virtual destructor included, other things
- IDD changes: n/a

### CODE REVIEW

**Comment by rraustad** [comment]
@Myoldmopar how do we get from this at line 2011 of HeatRecovery.cc:

ShowContinueError( "Average air volume flow rate is <50% or >130% of the nominal HX supply air volume flow rate." );

to this in the eio file:

```
Average air volume flow rate is &lt;50% or >130% of the nominal HX supply air volume flow rate.

and what's the fix? Switch "<" to "&lt;" ?
```

**Comment by rraustad** [comment]
@Myoldmopar I don't see the < or lt issue on my PC, it must be something to do with the CI

**Comment by EnergyArchmage** [comment]
I haven't been following this issue, (and there isn't much description) .   but it could be that you are really getting cycling air flow when it should be continuous.  That warning would kick in for low flow rates in an air-to-air HX. 

This line is using ContFanCycCoil when it should be CycFanCycCoil.

https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/HVACUnitarySystem.cc#L7210

also 
https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/HVACUnitarySystem.cc#L7239

**Comment by Myoldmopar** [comment]
So you are saying that currently the eio _does_ have the "& l t ;" string of characters?

**Comment by rraustad** [comment]
" & l t ; " only here in the eio for the failure DOAToWaterToAirHPInlet. My local PC shows <.

http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2016-05/3b72813e45ee6b849ac8bdb48f5e32208089ae1a-ffa000a74b8a3671244b176da0ee9f2f6bc6796f/DOAToWaterToAirHPInlet/Win64-Windows-7-VisualStudio-12/eplusout.err.diff.html

**Comment by rraustad** [comment]
@EnergyArchmage looks correct on the cont vs cyc fan lines in UnitarySystem code. PS, this issue is with water side temps and use of cycling vs constant water flow. The HX warning is separate.

**Comment by rraustad** [comment]
ZoneSysAvailManager is showing new warnings but I can't reproduce them in any CO weather file or using Chicago ??? Tried both debug and release versions.
Does CI run the example files with the exact weather file?

```
+   ** Warning ** GetSpecificHeatGlycol: Temperature is out of range (too low) for fluid [WATER] specific heat supplied values **
+   **   ~~~   ** ..Called From:CalcHPCoolingSimple:SourceSideInletTemp,Temperature=[-5.39], supplied data range=[0.00,125.00]
+   **   ~~~   **  Environment=DENVER CENTENNIAL ANN HTG 99.6% CONDNS DB, at Simulation time=12/21 00:00 - 00:15
```

**Comment by rraustad** [comment]
False alarm. Somehow I was looking at the previous push results. ZoneSysAvailManager has no diff's in the error file for the latest results.

**Comment by rraustad** [comment]
Summary of eio and err files:

```
EIO FILE:
All 7 files had differences in Warmup Convergence results.

ERROR FILE:
DOAToWaterToAirHPInlet
-   **   ~~~   ** Air volume flow rate ratio = 2.520E-002. Environment=CHICAGO_IL_USA ANNUAL COOLING 1% DESIGN CONDITIONS DB/MCWB, at Simulation time=07/21 05:30 - 05:45
+   **   ~~~   ** Air volume flow rate ratio = 2.518E-002. Environment=CHICAGO_IL_USA ANNUAL COOLING 1% DESIGN CONDITIONS DB/MCWB, at Simulation time=07/21 05:30 - 05:45

HeatPumpWaterToAirWithRHControl (these warnings now gone)
-   ** Warning ** GetSpecificHeatGlycol: Temperature is out of range (too high) for fluid [WATER] specific heat **
-   **   ~~~   ** ..Called From:PlantLoopSolver::UpdateAnyLoopDemandAlterations,Temperature=[176.28], supplied data range=[0.00,125.00]
-   **   ~~~   **  Environment=MIAMI INTL AP ANN CLG 1% CONDNS DB=>MWB, at Simulation time=07/21 02:10 - 02:20
    ************* Simulation Error Summary *************
-   *************
-   ************* ===== Recurring Error Summary =====
-   ************* The following recurring error messages occurred.
-   *************
-   *************  ** Warning ** GetSpecificHeatGlycol: Temperature out of range (too high) for fluid [WATER] specific heat **
-   *************  **   ~~~   **   This error occurred 30 total times;
-   *************  **   ~~~   **   during Warmup 0 times;
-   *************  **   ~~~   **   during Sizing 0 times.
-   *************  **   ~~~   **   Max=176.316177 {C}  Min=128.753602 {C}

HospitalLowEnergy (looks like when max iterations occur changed, also number of warnings increased)
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS5, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:05 - 00:10
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS4, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:05 - 00:10
    ** Warning ** CoolingTower:VariableSpeed "COOLSYS1 CHILLER TOWERSYS COOLTOWER" - Tower approach temperature is outside model boundaries at 0.00.
    **   ~~~   **  ...Valid range = 1.11 to 11.11. Occurrence info = CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:00 - 00:05
    **   ~~~   **  ...Tower approach temperature passed to the model = 0.00000
    **   ~~~   **  ...Approach temperatures outside model boundaries may not adversely affect tower performance.
    **   ~~~   **  ...This is not an unexpected occurrence when simulating actual conditions.
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS2, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:05 - 00:10
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:05 - 00:10
    ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:10 - 00:15
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS5, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:15 - 00:20
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:20 - 00:25
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS5, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:25 - 00:30
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:15 - 00:20
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS5, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:20 - 00:25
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS2, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:20 - 00:25
    ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS4, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:25 - 00:30
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:35 - 00:40
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:25 - 00:30
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS2, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:30 - 00:35
    ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:40 - 00:45
    ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:50 - 00:55
    ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS1, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:55 - 01:00
-   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS2, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:55 - 01:00
-   ** Warning ** CHILLER:ELECTRIC:REFORMULATEDEIR "CONDENSERLOOP CHILLER": The evaporator outlet temperature (31.05 C) is outside the range of evaporator outlet temperatures (X var) given in Cooling Capacity Function of Temperature biquadratic curve = CONDENSERLOOP CHILLER CLGCAPFUNCTEMPCURVE
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS5, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 00:55 - 01:00
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS4, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 01:00 - 01:05
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS4, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 01:05 - 01:10
+   ** Warning ** SolveAirLoopControllers: Maximum iterations (50) exceeded for DOAS2, at CHICAGO ANN CLG .4% CONDNS DB=>MWB, 07/21 01:10 - 01:15
+   ** Warning ** CHILLER:ELECTRIC:REFORMULATEDEIR "CONDENSERLOOP CHILLER": The evaporator outlet temperature (30.96 C) is outside the range of evaporator outlet temperatures (X var) given in Cooling Capacity Function of Temperature biquadratic curve = CONDENSERLOOP CHILLER CLGCAPFUNCTEMPCURVE
    **   ~~~   ** The range specified = 0.00 C to 20.00 C. Environment=CHICAGO ANN CLG .4% CONDNS DB=>MWB, at Simulation time=07/21 05:00 - 05:10
-   ** Warning ** CHILLER:ELECTRIC:REFORMULATEDEIR "CONDENSERLOOP CHILLER": The evaporator outlet temperature (31.05 C) is outside the range of evaporator outlet temperatures (X var) given in Electric Input to Cooling Output Ratio Function of Temperature biquadratic curve = CONDENSERLOOP CHILLER EIRFUNCTEMPCURVE
+   ** Warning ** CHILLER:ELECTRIC:REFORMULATEDEIR "CONDENSERLOOP CHILLER": The evaporator outlet temperature (30.96 C) is outside the range of evaporator outlet temperatures (X var) given in Electric Input to Cooling Output Ratio Function of Temperature biquadratic curve = CONDENSERLOOP CHILLER EIRFUNCTEMPCURVE
    **   ~~~   ** The range specified = 0.00 C to 20.00 C. Environment=CHICAGO ANN CLG .4% CONDNS DB=>MWB, at Simulation time=07/21 05:00 - 05:10
    ************* Simulation Error Summary *************
    ************* There are 7 unused schedules in input.
@@ -85,8 +85,8 @@
    *************  **   ~~~   **   during Sizing 0 times.
    *************  **   ~~~   **   Max=0.  Min=0.
    *************
-   *************  ** Warning ** SolveAirLoopControllers: Exceeding Maximum iterations for DOAS5 during CHICAGO ANN CLG .4% CONDNS DB=>MWB continues
-   *************  **   ~~~   **   This error occurred 123 total times;
+   *************  ** Warning ** SolveAirLoopControllers: Exceeding Maximum iterations for DOAS1 during CHICAGO ANN CLG .4% CONDNS DB=>MWB continues
+   *************  **   ~~~   **   This error occurred 108 total times;
    *************  **   ~~~   **   during Warmup 0 times;
    *************  **   ~~~   **   during Sizing 0 times.
    *************
@@ -94,16 +94,16 @@
    *************  **   ~~~   **   This error occurred 1078 total times;
    *************  **   ~~~   **   during Warmup 0 times;
    *************  **   ~~~   **   during Sizing 0 times.
-   *************  **   ~~~   **   Max=31.22296  Min=29.606023
+   *************  **   ~~~   **   Max=31.204736  Min=29.737826
    *************
    *************  ** Warning ** CHILLER:ELECTRIC:REFORMULATEDEIR "CONDENSERLOOP CHILLER": The evap outlet temp range in Electric Input to Cooling Output Ratio Function of Temp curve error continues.
    *************  **   ~~~   **   This error occurred 1078 total times;
    *************  **   ~~~   **   during Warmup 0 times;
    *************  **   ~~~   **   during Sizing 0 times.
-   *************  **   ~~~   **   Max=31.22296  Min=29.606023
+   *************  **   ~~~   **   Max=31.204736  Min=29.737826
    *************
-   *************  ** Warning ** SolveAirLoopControllers: Exceeding Maximum iterations for DOAS4 during CHICAGO ANN CLG .4% CONDNS WB=>MDB continues
-   *************  **   ~~~   **   This error occurred 340 total times;
+   *************  ** Warning ** SolveAirLoopControllers: Exceeding Maximum iterations for DOAS1 during CHICAGO ANN CLG .4% CONDNS WB=>MDB continues
+   *************  **   ~~~   **   This error occurred 381 total times;
```

File: src/EnergyPlus/WaterToAirHeatPumpSimple.cc [hunk 1]
```diff
@@ -825,16 +813,11 @@ namespace WaterToAirHeatPumpSimple {
...
-			SimpleWatertoAirHP( HPNum ).WaterMassFlowRate = SimpleWatertoAirHP( HPNum ).DesignWaterMassFlowRate * WaterPartLoad;
-
-			//    SimpleWatertoAirHP(HPNum)%WaterMassFlowRate =    SimpleWatertoAirHP(HPNum)%DesignWaterMassFlowRate
-
 			// Model requires the values to be calculated at full design flow rate for air and then scaled to part load ratio.
 			// So always start the calculations by setting the air flow rate to design flow rate.
 
-			//    SimpleWatertoAirHP(HPNum)%AirMassFlowRate   = Node(AirInletNode)%MassFlowRate
+			SimpleWatertoAirHP( HPNum ).WaterMassFlowRate = SimpleWatertoAirHP( HPNum ).DesignWaterMassFlowRate;
+
```
Comment (rraustad) [inline src/EnergyPlus/WaterToAirHeatPumpSimple.cc:824]: Underlying cause of issue. Water flow rate was calculated in proportion to RTF in Init (only for cycling water flow mode case). This caused 2 problems, 1) the full load capacity was calculated using the lowered water flow (in Calc), and 2) the outlet water temp would be excessive at low RTF.

File: src/EnergyPlus/WaterToAirHeatPumpSimple.cc [hunk 2]
```diff
@@ -1938,9 +1922,25 @@ namespace WaterToAirHeatPumpSimple {
 		SimpleWatertoAirHP( HPNum ).PartLoadRatio = PartLoadRatio;
 		SimpleWatertoAirHP( HPNum ).AirMassFlowRate = PLRCorrLoadSideMdot;
 
-		SimpleWatertoAirHP( HPNum ).WaterMassFlowRate = SourceSideMassFlowRate;
-		SimpleWatertoAirHP( HPNum ).OutletWaterTemp = SourceSideInletTemp + QSource / ( SourceSideMassFlowRate * CpWater );
-		SimpleWatertoAirHP( HPNum ).OutletWaterEnthalpy = SourceSideInletEnth + QSource / SourceSideMassFlowRate;
+		if ( ( SimpleWatertoAirHP( HPNum ).WaterCyclingMode ) == WaterCycling ) {
+			SimpleWatertoAirHP( HPNum ).WaterMassFlowRate = SourceSideMassFlowRate * PartLoadRatio;
+			SetComponentFlowRate( SimpleWatertoAirHP( HPNum ).WaterMassFlowRate, SimpleWatertoAirHP( HPNum ).WaterInletNodeNum, SimpleWatertoAirHP( HPNum ).WaterOutletNodeNum, SimpleWatertoAirHP( HPNum ).LoopNum, SimpleWatertoAirHP( HPNum ).LoopSide, SimpleWatertoAirHP( HPNum ).BranchNum, SimpleWatertoAirHP( HPNum ).CompNum );
```
Comment (rraustad) [inline src/EnergyPlus/WaterToAirHeatPumpSimple.cc:1931]: Set the new water side flow rate. This is the average water flow over the time step and does not affect full load performance calculated above. For cycling models, the fluid flow is averaged over the time step and outlet node properties denote full capacity. For constant models, the fluid flow is full flow over the time step and outlet node properties are averaged over the time step.

**Comment by nrel-bot-3** [comment]
@Myoldmopar @rraustad @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@Myoldmopar @rraustad @lgentile it has been 14 days since this pull request was last updated.

**Comment by mjwitte** [comment]
@rraustad I've merged in current develop, but the unit test fails in SetComponentFlowRate with loopside = 0 called from [here](https://github.com/NREL/EnergyPlus/blob/ef760a47c310dfe425d3555d2ed79c1e27d293f3/src/EnergyPlus/WaterToAirHeatPumpSimple.cc#L865) when running all unit tests locally.  Probably a clear_state problem.

**Comment by rraustad** [comment]
I can't explain the lack of high delta T at low Mdot, but the point of this change was to make the water side delta T constant while modulating the water flow rate. Cycling system has full load outlet Temp with flow rate proportional to PLR. Constant system has full load water flow rate with delta T proportional to PLR.

**Comment by rraustad** [comment]
I ran the full unit test suite before pulling develop and the unit test is not failing. I'll have to pull develop to see if that has an impact.

**Comment by rraustad** [comment]
Not sure how this could have worked without the clear_state.

**Comment by mjwitte** [comment]
Found the original defect file (added to EnergyPlusDevSupport for reference), reproduced the original problem, and confirmed that this fix eliminates the glycol temperature warnings. Other diffs look reasonable.

**Comment by rraustad** [comment]
Whew! Thanks for the follow up.

Sent from my Sprint Samsung Galaxy S7.
-------- Original message --------From: "Michael J. Witte" notifications@github.com Date: 7/14/16  8:00 PM  (GMT-05:00) To: NREL/EnergyPlus EnergyPlus@noreply.github.com Cc: Richard Raustad rraustad@fsec.ucf.edu, Mention mention@noreply.github.com Subject: Re: [NREL/EnergyPlus] Corrections to WaterToAirHeatPump (WSHP) water
  flow control (#5691) 
Found the original defect file (added to EnergyPlusDevSupport for reference), reproduced the original problem, and confirmed that this fix eliminates the glycol temperature warnings. Other diffs look reasonable.

**Pull request merged**